### ISSUE TITLE
[Feature Request] Postponed evaluation of non-self-referential type hints, Part II: "Ignore PEP 563 for the win."

### ISSUE THREAD

**Issue by leycec** [beartype/beartype#233]
Typing behemoth @st-- has [sadly uncovered in a related issue that @beartype appears to be incompatible with the third-party `jaxtyping.install_import_hook()` import hook when PEP 563 (i.e., `from __future__ import annotations`) is enabled for modules subsequently imported by that import hook](https://github.com/beartype/beartype/issues/226#issuecomment-1511019847).

Take it away, @st--!

> I'm encountering a similar issue, but in a different setup, where the exception is `beartype.roar.BeartypeDecorHintForwardRefException: Forward reference "gpjax.kernels.approximations.rff.Optional[List[int]]" syntactically invalid as module attribute name.` and the issue is a base class in one file

> It's the from `__future__ import` annotations. MRE:

> File `mre/__init__.py`:

```
from jaxtyping import install_import_hook

with install_import_hook("mre", "beartype.beartype"):
    from .base import *
    from .rff import *
```

> File `mre/base.py`:

```python
from __future__ import annotations

from beartype.typing import List, Optional
from dataclasses import dataclass

@dataclass
class AbstractKernel:
    """Base kernel class."""

    active_dims: Optional[List[int]] = None
```

> File `mre/rff.py`:

```python
#from __future__ import annotations
from dataclasses import dataclass

from .base import AbstractKernel

@dataclass
class RFF(AbstractKernel):
    pass
```

> Then `import mre` gives the following error:

```
BeartypeDecorHintForwardRefException: Forward reference "mre2.rff.Optional[List[int]]" syntactically invalid as module attribute name.
```

> If you comment out the from `__future__ import annotations` in the `base.py`, the error disappears.

:scream:

**Comment by leycec** [comment]
**Fascinating...** and horrifying, @st--! First, a recap. What's happening here is that @beartype is failing to detect that your `AbstractKernel` class has been declared under PEP 563 (i.e., `from __future__ import annotations`). But that should *never* happen. Tests say that never happens. Moreover, that's never happened before. Which can only mean one thing...

The third-party `jaxtyping` import hook (i.e., `jaxtyping.install_import_hook()`) is *probably* to blame, somehow. To decide whether a class has been declared under PEP 563, @beartype just introspects the module declaring that class for a global attribute named `annotations` whose value is the `__future__.annotations` singleton. Since this trivially works otherwise, the `jaxtyping` import hook is *probably* accidentally interfering in that introspection – like modifying the `__module__` attribute of your `AbstractKernel` class to be something *other* than the string `"mre.base"` (i.e., the fully-qualified name of your submodule declaring that class), maybe. Possibly. Not even Dagon the fire-breathing dragon knows anymore. :fire:  :dragon_face:

Give me a day or three and I'll investigate deeply on behalf of humanity and QA. Until then, I only have horrible parental advice that you should consider ignoring (just like with real parents). Consider either:

* Avoid calling the `jaxtyping.install_import_hook()` context manager for a bit. Instead, please manually decorate literally the entire world with `@beartype`. Look. I said it was horrible parental advice, didn't I? It is. Nobody wants to manually decorate literally the entire world with `@beartype`. Still, this isn't the worst possible choice. `beartype 0.14.0` (to be released before the fiery death of the Sun) will ship its own shiny import hook that will *probably* integrate much better with PEP 563. I promise the world and deliver much less.
* **(Recommended)** Comment out all `from __future__ import annotations` statements. PEP 563 is pretty much broken at runtime, anyway. It simply doesn't work; even when it *does* work, it harms runtime performance. Even though @beartype ostensibly has the best PEP 563 support of any existing runtime type-checker, there's only so much @beartype can do here. PEP 563 lacks a runtime API despite breaking Python at runtime. Dagon, I summon thee to purge `from __future__ import annotations` from this poor victim's codebase! :fire:  :dragon_face:

**Comment by patrick-kidger** [comment]
Just saw this. FWIW the `jaxtyping.install_import_hook` adds a couple of decorators (`jaxtyping.jaxtyped` and `beartype.beartype`) to every function definition in the AST of the module. LMK if jaxtyping could do something more here.

---

I would note that beartype inspecting `__module__` is pretty fragile. It's common to re-assign the `__module__` attribute so that e.g. `str(Foo)` produces `<class package.Foo>` rather than `<class package._src.foo.Foo>`; e.g. https://github.com/google/jax/pull/13270.

Just `isinstance(f.__annotations__["arg"], str)` may work instead?

**Comment by leycec** [comment]
> I would note that beartype inspecting `__module__` is pretty fragile.

You are not wrong. You never are! This is why I pound sand in mute frustration whenever someone utters the two worst words in the English lexicon: "PEP. 563."

> It's common to re-assign the `__module__` attribute...

*...heh.* Yup. @beartype itself routinely does this, which only goes to show that trust is risky. Hypocrisy and shamelessness are common bedfellows here on GitHub.

> Just `isinstance(f.__annotations__["arg"], str)` may work instead?

*...heh.* I wish. Wouldn't that have been delightful! Sadly, forward references are usually just strings. Sometimes, they're actually `typing.ForwardRef` instances – because some but not all `typing` constructs internally coerce strings into `typing.ForwardRef` instances, for unknown reasons. Usually, though, they're just strings.

In other words, forward references and PEP 563 syntactically collide. This should have been obvious to everyone who authored and reviewed PEP 563. It wasn't. Now, we tentatively pick up the broken shards of the Python parser by (*...wait for it*) inspecting `__module__`. :face_exhaling: 

PEP 563. It's constitutionally broken, which is a deeper shade of broken than most broken things. We can't fix broken; we can only pretend to support it.

**Comment by leycec** [comment]
Actually... *you're right*. Our existing PEP 563 support is awful. Technically, we're just doing what PEP 563 says to do – but PEP 563 is broken, so nobody should do what PEP 563 says to do.

Instead, @beartype should do what @TeamSpen210 [ingeniously said to do](https://github.com/beartype/beartype/issues/226#issuecomment-1484045900) at #226:

> Instead of writing the parser manually, there's a simpler approach, somewhat described [here for PEP 649](https://discuss.python.org/t/two-polls-on-how-to-revise-pep-649/23628) (Deferred Evaluation Of Annotations Using Descriptors). Instead of executing the annotation normally, instead a custom dict subclass could be passed as the locals dictionary. That allows intercepting all namespace lookups. So we could do a normal lookup of globals/locals, but if a lookup fails, create and inject a proxy object which has all the annotation-relevant special methods defined. Or even unconditionally return a proxy for every name, perhaps to construct a tree of door objects instead.

Let's paraphrase @TeamSpen210. His brilliance illuminates the dark corners. First, we provide full-blown support for arbitrarily complex stringified type hints (e.g., `def muh_func(muh_arg: "Union[List[MyCls], MyCls]]"): ...`) via dictionary trickery. Then, we throw out our existing PEP 563 support altogether. Last, we ignore PEP 563 for the remainder of our days.

Solving #226 should thus solve this. In theory. Maybe. Let's rename this issue accordingly.

**Comment by leycec** [comment]
**Ah-hah!** Resolving #226 does, indeed, appear to have perfectly resolved this issue as well. I am shocked face. :scream: 

Let's quietly close this and pretend PEP 563 never happened, everybody.
